README.md
# Clocking Timer

[![Version](https://img.shields.io/badge/version-1.0.0-blue.svg)](https://marketplace.visualstudio.com/items?itemName=matbel91765.clockingtimer)

## Description

Cette extension VSCode permet de suivre le temps passé sur un projet. Elle offre des fonctionnalités telles que le démarrage/arrêt automatique du timer et l'affichage du temps total passé.

## Installation

Pour installer l'extension, accédez au [VSCode Marketplace](https://marketplace.visualstudio.com/items?itemName=matbel91765.clockingtimer) et cliquez sur "Installer".

## Utilisation

Pour utiliser l'extension, cliquez sur le bouton "Start Timer" dans la barre d'état pour commencer à suivre le temps. Le bouton changera en "Stop Timer" une fois activé.

![Utilisation](https://github.com/matbel91765/clockingtimer/blob/master/res/time-bar.png?raw=true)

## Fonctionnalités

- **Timer Automatique**: Démarre et arrête automatiquement le timer en fonction de l'activité dans l'éditeur.
- **Affichage du Temps Total**: Montre le temps total passé sur le projet.

## Contribuer

Les contributions sont toujours les bienvenues! Veuillez forker le repo et créer une pull request avec vos modifications.

## Support et Feedback

Pour toute question ou suggestion, ouvrez une issue sur [GitHub](https://github.com/matbel91765/clockingtimer/issues).

## Auteurs et Remerciements

Créé par Belouar-Pruvot Mathis

## Licence

Ce projet est sous licence MIT. Voir le fichier [LICENSE](https://github.com/matbel91765/clockingtimer/blob/master/LICENSE) pour plus de détails.


CHANGELOG.md
# Change Log

All notable changes to the "clockingtimer" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/main.ts
import * as vscode from 'vscode';
import { TimerController } from './timerController';
import { UIComponents } from './uiComponents';
import { handleUserActivity, handleWorkspaceChange } from './eventHandlers';

let timerController: TimerController;
let uiComponents: UIComponents;

export function activate(context: vscode.ExtensionContext) {
    console.log('Votre extension "clockingtimer" est active.');

    // Création timerController sans uiComponents
    timerController = new TimerController(context);

    // Création uiComponents avec timerController
    uiComponents = new UIComponents(context, timerController);

    // Mise à jour du timerController avec uiComponents
    timerController.setUIComponents(uiComponents);

    context.subscriptions.push(vscode.commands.registerCommand('clockingtimer.toggleTimer', () => uiComponents.toggleTimer()));
    context.subscriptions.push(vscode.commands.registerCommand('clockingtimer.resetTimer', () => uiComponents.resetTimer()));

    vscode.workspace.onDidChangeTextDocument(() => handleUserActivity(timerController), null, context.subscriptions);
    vscode.window.onDidChangeActiveTextEditor(() => handleUserActivity(timerController), null, context.subscriptions);
    vscode.workspace.onDidChangeWorkspaceFolders((e) => handleWorkspaceChange(timerController));

    timerController.restorePreviousSession();
}

export function deactivate(context: vscode.ExtensionContext) {
    if (timerController) {
        timerController.dispose();
    }
    if (uiComponents) {
        uiComponents.dispose();
    }
}


webpack.config.js
//@ts-check

'use strict';

const path = require('path');

//@ts-check
/** @typedef {import('webpack').Configuration} WebpackConfig **/

/** @type WebpackConfig */
const extensionConfig = {
  target: 'node', // VS Code extensions run in a Node.js-context 📖 -> https://webpack.js.org/configuration/node/
	mode: 'none', // this leaves the source code as close as possible to the original (when packaging we set this to 'production')

  entry: './src/main.ts', // the entry point of this extension, 📖 -> https://webpack.js.org/configuration/entry-context/
  output: {
    // the bundle is stored in the 'dist' folder (check package.json), 📖 -> https://webpack.js.org/configuration/output/
    path: path.resolve(__dirname, 'dist'),
    filename: 'extension.js',
    libraryTarget: 'commonjs2'
  },
  externals: {
    vscode: 'commonjs vscode' // the vscode-module is created on-the-fly and must be excluded. Add other modules that cannot be webpack'ed, 📖 -> https://webpack.js.org/configuration/externals/
    // modules added here also need to be added in the .vscodeignore file
  },
  resolve: {
    // support reading TypeScript and JavaScript files, 📖 -> https://github.com/TypeStrong/ts-loader
    extensions: ['.ts', '.js']
  },
  module: {
    rules: [
      {
        test: /\.ts$/,
        exclude: /node_modules/,
        use: [
          {
            loader: 'ts-loader'
          }
        ]
      }
    ]
  },
  devtool: 'nosources-source-map',
  infrastructureLogging: {
    level: "log", // enables logging required for problem matchers
  },
};
module.exports = [ extensionConfig ];

src/uiComponents.ts
import * as vscode from 'vscode';
import { TimerController } from './timerController';

export class UIComponents {
    private timerButton: vscode.StatusBarItem;
    private timerDisplay: vscode.StatusBarItem;
    private totalTimeDisplay: vscode.StatusBarItem;

    constructor(private context: vscode.ExtensionContext, private timerController: TimerController) {
        this.timerButton = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right, 98);
        this.timerButton.command = 'clockingtimer.toggleTimer';
        this.timerButton.show();

        this.timerDisplay = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right, 99);
        this.timerDisplay.show();

        this.totalTimeDisplay = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Left, 1); 
        this.totalTimeDisplay.text = `Total Time: ${this.formatElapsedTime(this.timerController.totalElapsedTime)}`;
        this.totalTimeDisplay.show();

        // Mettre à jour l'UI selon l'état initial du TimerController
        this.updateUI();
    }

    toggleTimer() {
        this.timerController.toggleTimer();
        this.updateUI();
    }

    resetTimer() {
        this.timerController.resetTimer();
        this.updateUI();
    }

    dispose() {
        if (this.timerButton) {
            this.timerButton.dispose();
        }
        if (this.timerDisplay) {
            this.timerDisplay.dispose();
        }
        if (this.totalTimeDisplay) {
            this.totalTimeDisplay.dispose();
        }
    }

    public handleTimerStateChange() {
        this.updateUI();
    }

    public updateTimerDisplay(elapsedMillis: number) {
        this.timerDisplay.text = `Timer: ${this.formatElapsedTime(elapsedMillis)}`;
    }

    private updateUI() {
        if (this.timerController.isTimerRunning) {
            this.timerButton.text = `$(clock) Stop Timer`;
        } else {
            this.timerButton.text = `$(clock) Start Timer`;
            // Mettez à jour le timerDisplay pour montrer le temps écoulé jusqu'à l'arrêt
            this.timerDisplay.text = `Timer: ${this.formatElapsedTime(this.timerController.elapsedSessionTime)}`;
        }
        // Mettre à jour totalTimeDisplay indépendamment de l'état du chronomètre
        this.totalTimeDisplay.text = `Total Time: ${this.formatElapsedTime(this.timerController.totalElapsedTime)}`;
    }
    

    private formatElapsedTime(elapsedMillis: number): string {
        let seconds = Math.floor(elapsedMillis / 1000);
        let minutes = Math.floor(seconds / 60);
        seconds %= 60;
        let hours = Math.floor(minutes / 60);
        minutes %= 60;
        return `${hours}h ${minutes}m ${seconds}s`;
    }
}


src/eventHandlers.ts
import { TimerController } from './timerController';

let activityTimeout: NodeJS.Timeout | undefined;

export function handleUserActivity(timerController: TimerController) {
    if (!timerController.isTimerRunning) {
        timerController.startTimer();
    }

    clearTimeout(activityTimeout);
    activityTimeout = setTimeout(() => {
        if (timerController.isTimerRunning) {
            timerController.stopTimer();
        }
    }, 30000); // 30 secondes d'inactivité
}

export function handleWorkspaceChange(timerController: TimerController) {
    if (timerController.isTimerRunning && timerController.startTime) {
        let sessionTime = new Date().getTime() - timerController.startTime.getTime();
        timerController.elapsedSessionTime += sessionTime;
        timerController.totalElapsedTime += sessionTime;
        
        // Utilisation du getter pour accéder à context
        timerController.vscodeContext.workspaceState.update('elapsedTime', timerController.elapsedSessionTime);
        timerController.vscodeContext.workspaceState.update('totalElapsedTime', timerController.totalElapsedTime);
    }
}


src/timerController.ts
import * as vscode from 'vscode';
import { UIComponents } from './uiComponents';

export class TimerController {
    private _startTime: Date | undefined;
    private _timerInterval: NodeJS.Timer | undefined;
    private _elapsedSessionTime: number = 0;
    private _totalElapsedTime: number;
    private _isTimerRunning: boolean = false;
    private uiComponents?: UIComponents;

    constructor(private context: vscode.ExtensionContext, uiComponents?: UIComponents) {
        if (uiComponents) {
            this.uiComponents = uiComponents;
        }
        this._totalElapsedTime = context.workspaceState.get<number>('totalElapsedTime') || 0;
    }

    get isTimerRunning(): boolean {
        return this._isTimerRunning;
    }

    get startTime(): Date | undefined {
        return this._startTime;
    }

    get vscodeContext(): vscode.ExtensionContext {
        return this.context;
    }

    get elapsedSessionTime(): number {
        return this._elapsedSessionTime;
    }

    set elapsedSessionTime(value: number) {
        this._elapsedSessionTime = value;
    }

    get totalElapsedTime(): number {
        return this._totalElapsedTime;
    }

    set totalElapsedTime(value: number) {
        this._totalElapsedTime = value;
    }

    toggleTimer() {
        if (this._isTimerRunning) {
            this.stopTimer();
        } else {
            this.startTimer();
        }
    }

    startTimer() {
        if (!this._isTimerRunning) {
            this._startTime = new Date();
        }
        this._timerInterval = setInterval(() => this.updateTimer(), 1000);
        this._isTimerRunning = true;
        if (this.uiComponents) {
            this.uiComponents.handleTimerStateChange();
        }
    }

    stopTimer() {
        if (this._isTimerRunning && this._startTime) {
            clearInterval(this._timerInterval as NodeJS.Timeout);
            this._timerInterval = undefined;

            let sessionTime = new Date().getTime() - this._startTime.getTime();
            this._elapsedSessionTime += sessionTime;
            this._totalElapsedTime += sessionTime;
            this.context.workspaceState.update('elapsedTime', this._elapsedSessionTime);
            this.context.workspaceState.update('totalElapsedTime', this._totalElapsedTime);

            this._isTimerRunning = false;
        }
        if (this.uiComponents) {
            this.uiComponents.handleTimerStateChange();
        }
    }

    dispose() {
        if (this._timerInterval) {
            clearInterval(this._timerInterval as NodeJS.Timeout);
            this._timerInterval = undefined;
        }
        // Toute autre nettoyage nécessaire
    }

    resetTimer() {
        clearInterval(this._timerInterval as NodeJS.Timeout);
        this._timerInterval = undefined;
        this._startTime = undefined;
        this._elapsedSessionTime = 0;
        this._totalElapsedTime = 0;
        this._isTimerRunning = false;

        this.context.workspaceState.update('elapsedTime', undefined);
        this.context.workspaceState.update('totalElapsedTime', undefined);
        if (this.uiComponents) {
            this.uiComponents.handleTimerStateChange();
        }
    }

    private updateTimer() {
        if (this._startTime) {
            const currentTime = new Date();
            const elapsedMillis = currentTime.getTime() - this._startTime.getTime() + this._elapsedSessionTime;
            if (this.uiComponents) {
                this.uiComponents.updateTimerDisplay(elapsedMillis);
            }
        }
    }

    restorePreviousSession() {
        let savedTime = this.context.workspaceState.get<number>('elapsedTime');
        if (savedTime && savedTime > 0) {
            this._startTime = new Date(Date.now() - savedTime);
            this.startTimer();
        }
    }

    setUIComponents(uiComponents: UIComponents) {
        this.uiComponents = uiComponents;
    }
}


src/extension.ts
import * as vscode from 'vscode';

let startTime: Date | undefined;
let isTimerRunning: boolean = false;
let timerButton: vscode.StatusBarItem;
let timerDisplay: vscode.StatusBarItem;
let timerInterval: NodeJS.Timer | undefined;
let extensionContext: vscode.ExtensionContext;
let activityTimeout: NodeJS.Timeout | undefined;
let totalTimeDisplay: vscode.StatusBarItem;
let totalElapsedTime: number = 0;
let elapsedSessionTime: number = 0;

export function activate(context: vscode.ExtensionContext) {
    extensionContext = context;
    console.log('Votre extension "clockingtimer" est active.');

    // Créer le bouton "Start Timer"
    timerButton = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right, 98);
    timerButton.command = 'clockingtimer.toggleTimer';
    timerButton.show();

    // Créer l'affichage du chronomètre
    timerDisplay = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right, 99);
    timerDisplay.show();

    // Créer l'affichage du temps total
    totalElapsedTime = context.workspaceState.get<number>('totalElapsedTime') || 0;
    totalTimeDisplay = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Left, 1); 
    totalTimeDisplay.text = `Total Time: ${formatElapsedTime(totalElapsedTime)}`;
    totalTimeDisplay.show();

    let savedTime = context.workspaceState.get<number>('elapsedTime');
    if (savedTime && savedTime > 0) {
        startTime = new Date(Date.now() - savedTime);
        startTimer();
    } else {
        timerButton.text = `$(clock) Start Timer`;
        timerDisplay.text = `Timer: 00h 00m 00s`;
    }

    let toggleTimerCommand = vscode.commands.registerCommand('clockingtimer.toggleTimer', toggleTimer);
    context.subscriptions.push(toggleTimerCommand, timerButton, timerDisplay);

    let resetTimerCommand = vscode.commands.registerCommand('clockingtimer.resetTimer', resetTimer);
    context.subscriptions.push(resetTimerCommand);

    vscode.workspace.onDidChangeTextDocument(onUserActivity, null, context.subscriptions);
    vscode.window.onDidChangeActiveTextEditor(onUserActivity, null, context.subscriptions);

    context.subscriptions.push(vscode.workspace.onDidChangeWorkspaceFolders((e) => {
        handleWorkspaceChange();
    }));
}

function toggleTimer() {
    if (isTimerRunning) {
        stopTimer();
        timerButton.text = `$(clock) Start Timer`;
    } else {
        startTimer();
        timerButton.text = `$(clock) Stop Timer`;
    }
}

function startTimer() {
    if (!isTimerRunning) {
        startTime = new Date();
        timerInterval = setInterval(updateTimerDisplay, 1000);
        timerButton.text = `$(clock) Stop Timer`;
        isTimerRunning = true;
    }
}

function stopTimer() {
    if (isTimerRunning && startTime) {
        clearInterval(timerInterval as NodeJS.Timeout);
        timerInterval = undefined;

        let sessionTime = new Date().getTime() - startTime.getTime();
        elapsedSessionTime += sessionTime;
        totalElapsedTime += sessionTime;
        extensionContext.workspaceState.update('elapsedTime', elapsedSessionTime);
        extensionContext.workspaceState.update('totalElapsedTime', totalElapsedTime);

        totalTimeDisplay.text = `Total Time: ${formatElapsedTime(totalElapsedTime)}`;

        timerButton.text = `$(clock) Start Timer`;
        isTimerRunning = false;
    }
}

function resetTimer() {
    clearInterval(timerInterval as NodeJS.Timeout);
    timerInterval = undefined;
    startTime = undefined;
    elapsedSessionTime = 0;
    totalElapsedTime = 0;
    isTimerRunning = false;

    timerDisplay.text = `Timer: 00h 00m 00s`;
    totalTimeDisplay.text = `Total Time: 00h 00m 00s`;
    timerButton.text = `$(clock) Start Timer`;

    extensionContext.workspaceState.update('elapsedTime', undefined);
    extensionContext.workspaceState.update('totalElapsedTime', undefined);
}

function updateTimerDisplay() {
    if (startTime) {
        const currentTime = new Date();
        const elapsedMillis = currentTime.getTime() - startTime.getTime() + elapsedSessionTime;
        timerDisplay.text = `Timer: ${formatElapsedTime(elapsedMillis)}`;
    }
}

function onUserActivity() {
    if (!isTimerRunning) {
        startTimer();
    }

    clearTimeout(activityTimeout);
    activityTimeout = setTimeout(() => {
        if (isTimerRunning) {
            stopTimer();
        }
    }, 30000); // 30 secondes d'inactivité
}

function formatElapsedTime(elapsedMillis: number): string {
    let seconds = Math.floor(elapsedMillis / 1000);
    let minutes = Math.floor(seconds / 60);
    seconds %= 60;
    let hours = Math.floor(minutes / 60);
    minutes %= 60;
    return `${hours}h ${minutes}m ${seconds}s`;
}

function handleWorkspaceChange() {
    if (isTimerRunning && startTime) {
        let sessionTime = new Date().getTime() - startTime.getTime();
        elapsedSessionTime += sessionTime;
        totalElapsedTime += sessionTime;
        extensionContext.workspaceState.update('elapsedTime', elapsedSessionTime);
        extensionContext.workspaceState.update('totalElapsedTime', totalElapsedTime);
    }
}

export function deactivate() {
    if (isTimerRunning && startTime) {
        let sessionTime = new Date().getTime() - startTime.getTime();
        elapsedSessionTime += sessionTime;
        totalElapsedTime += sessionTime;
        extensionContext.workspaceState.update('elapsedTime', elapsedSessionTime);
        extensionContext.workspaceState.update('totalElapsedTime', totalElapsedTime);
    }

    if (timerInterval) {
        clearInterval(timerInterval as NodeJS.Timeout);
    }
    if (timerButton) {
        timerButton.dispose();
    }
    if (timerDisplay) {
        timerDisplay.dispose();
    }
    if (totalTimeDisplay) {
        totalTimeDisplay.dispose();
    }
}
